File: dqoai/src/main/java/ai/dqo/rest/models/comparison/CompareThresholdsModel.java
/*
 * Copyright © 2021 DQO.ai (support@dqo.ai)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ai.dqo.rest.models.comparison;

import ai.dqo.checks.comparison.ComparisonCheckRules;
import ai.dqo.rules.comparison.MaxDiffPercentRule0ParametersSpec;
import ai.dqo.rules.comparison.MaxDiffPercentRule1ParametersSpec;
import ai.dqo.rules.comparison.MaxDiffPercentRule5ParametersSpec;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

/**
 * Model with the custom compare threshold levels for raising data quality issues at different severity levels
 * when the difference between the compared (tested) table and the reference table (the source of truth) exceed given
 * thresholds as a percentage of difference between the actual value and the expected value from the reference table.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "CompareThresholdsModel", description = "Model with the compare threshold levels for raising data quality issues at different severity levels " +
        "when the difference between the compared (tested) table and the reference table (the source of truth) exceed given " +
        "thresholds as a percentage of difference between the actual value and the expected value from the reference table.")
@Data
public class CompareThresholdsModel {
    @JsonPropertyDescription("Enables raising a warning severity data quality issue when the difference between the measure on the compared table and the reference table is above the warning level threshold (the percentage difference).")
    private boolean enableWarning = true;

    @JsonPropertyDescription("The percentage difference between the measure value on the compared table and the reference table that raises a warning severity data quality issue when the difference is bigger.")
    private double warningDifferencePercent = 0.0;

    @JsonPropertyDescription("Enables raising an error severity data quality issue when the difference between the measure on the compared table and the reference table is above the error level threshold (the percentage difference).")
    private boolean enableError = true;

    @JsonPropertyDescription("The percentage difference between the measure value on the compared table and the reference table that raises an error severity data quality issue when the difference is bigger.")
    private double errorDifferencePercent = 1.0;

    @JsonPropertyDescription("Enables raising a fatal severity data quality issue when the difference between the measure on the compared table and the reference table is above the fatal level threshold (the percentage difference).")
    private boolean enableFatal = true;

    @JsonPropertyDescription("The percentage difference between the measure value on the compared table and the reference table that raises a fatal severity data quality issue when the difference is bigger.")
    private double fatalDifferencePercent = 5.0;

    /**
     * Creates a compare threshold object from the comparison check.
     * @param comparisonCheckRules Comparison check specification.
     * @return Compare threshold model or null when the check was also null.
     */
    public static CompareThresholdsModel fromComparisonCheckSpec(ComparisonCheckRules comparisonCheckRules) {
        if (comparisonCheckRules == null) {
            return null;
        }

        CompareThresholdsModel thresholdsModel = new CompareThresholdsModel();

        MaxDiffPercentRule0ParametersSpec warning = comparisonCheckRules.getWarning();
        thresholdsModel.enableWarning = warning != null;
        if (warning != null) {
            thresholdsModel.warningDifferencePercent = warning.getMaxDiffPercent();
        }

        MaxDiffPercentRule1ParametersSpec error = comparisonCheckRules.getError();
        thresholdsModel.enableError = error != null;
        if (error != null) {
            thresholdsModel.errorDifferencePercent = error.getMaxDiffPercent();
        }

        MaxDiffPercentRule5ParametersSpec fatal = comparisonCheckRules.getFatal();
        thresholdsModel.enableFatal = fatal != null;
        if (fatal != null) {
            thresholdsModel.fatalDifferencePercent = fatal.getMaxDiffPercent();
        }

        return thresholdsModel;
    }

    /**
     * Copies the configuration of comparison difference thresholds to the comparison rules, creating or removing rules that are enabled or disabled in the threshold model.
     * @param targetComparisonCheckSpec Target comparison check specification to alter the rules.
     */
    public void copyToComparisonCheckSpec(ComparisonCheckRules targetComparisonCheckSpec) {
        targetComparisonCheckSpec.setWarning(this.enableWarning ? new MaxDiffPercentRule0ParametersSpec(this.warningDifferencePercent) : null);
        targetComparisonCheckSpec.setError(this.enableError ? new MaxDiffPercentRule1ParametersSpec(this.errorDifferencePercent) : null);
        targetComparisonCheckSpec.setFatal(this.enableFatal ? new MaxDiffPercentRule5ParametersSpec(this.fatalDifferencePercent) : null);
    }

    /**
     * Checks if any rule is enabled, so the check should be set.
     * @return True when at least one threshold level (warning, error, fatal) is enabled. False when all rules are disabled, so the check should be disabled.
     */
    @JsonIgnore
    public boolean isAnyThresholdRuleEnabled() {
        return this.enableWarning || this.enableError || this.enableFatal;
    }
}

File: distribution/python/dqops/client/models/compare_thresholds_model.py
from typing import Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

T = TypeVar("T", bound="CompareThresholdsModel")


@_attrs_define
class CompareThresholdsModel:
    """Model with the compare threshold levels for raising data quality issues at different severity levels when the
    difference between the compared (tested) table and the reference table (the source of truth) exceed given thresholds
    as a percentage of difference between the actual value and the expected value from the reference table.

        Attributes:
            warning_difference_percent (Union[Unset, float]): The percentage difference between the measure value on the
                compared table and the reference table that raises a warning severity data quality issue when the difference is
                bigger.
            error_difference_percent (Union[Unset, float]): The percentage difference between the measure value on the
                compared table and the reference table that raises an error severity data quality issue when the difference is
                bigger.
            fatal_difference_percent (Union[Unset, float]): The percentage difference between the measure value on the
                compared table and the reference table that raises a fatal severity data quality issue when the difference is
                bigger.
    """

    warning_difference_percent: Union[Unset, float] = UNSET
    error_difference_percent: Union[Unset, float] = UNSET
    fatal_difference_percent: Union[Unset, float] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        warning_difference_percent = self.warning_difference_percent
        error_difference_percent = self.error_difference_percent
        fatal_difference_percent = self.fatal_difference_percent

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if warning_difference_percent is not UNSET:
            field_dict["warning_difference_percent"] = warning_difference_percent
        if error_difference_percent is not UNSET:
            field_dict["error_difference_percent"] = error_difference_percent
        if fatal_difference_percent is not UNSET:
            field_dict["fatal_difference_percent"] = fatal_difference_percent

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        d = src_dict.copy()
        warning_difference_percent = d.pop("warning_difference_percent", UNSET)

        error_difference_percent = d.pop("error_difference_percent", UNSET)

        fatal_difference_percent = d.pop("fatal_difference_percent", UNSET)

        compare_thresholds_model = cls(
            warning_difference_percent=warning_difference_percent,
            error_difference_percent=error_difference_percent,
            fatal_difference_percent=fatal_difference_percent,
        )

        compare_thresholds_model.additional_properties = d
        return compare_thresholds_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
